File: out_go_v5/migrate/clean.go
package refactorbot

import (
	"bytes"
	"fmt"
	"strings"
)

type CodeSmell struct {
	Name        string
	Description string
	LineHint    *int
}

type CleanupEngine struct {
	Model CodexModel
}

type CodexModel interface {
	Complete(prompt string) (string, error)
}

func NewCleanupEngine(model CodexModel) *CleanupEngine {
	return &CleanupEngine{Model: model}
}

func (c *CleanupEngine) DetectSmells(code string, lang Language) []*CodeSmell {
	var smells []*CodeSmell

	switch lang {
	case Python:
		// Smell: JSON dump of objects
		if strings.Contains(code, "json.dump(") && !strings.Contains(code, "asdict(") && !strings.Contains(code, "JSONEncoder") {
			smells = append(smells, &CodeSmell{
				Name: "Custom Object Serialization",
				Description: "Found `json.dump` without `asdict` or custom encoder. This will crash for custom classes.",
				LineHint:     nil,
			})
		}
		// Smell: Missing dataclasses
		if strings.Contains(code, "def __init__(self") && !strings.Contains(code, "@dataclass") {
			smells = append(smells, &CodeSmell{
				Name: "Missing Dataclasses",
				Description: "Boilerplate __init__ mirrors source struct; consider using @dataclass.",
				LineHint:     nil,
			})
		}
		// Smell: Static methods for defaults
		if strings.Contains(code, "@staticmethod") && (strings.Contains(code, "default") || strings.Contains(code, "max_entries")) {
			smells = append(smells, &CodeSmell{
				Name: "Rust-style Static Helpers",
				Description: "Static methods like `default_max_entries` should be converted to default parameters.",
				LineHint:     nil,
			})
		}
	}

	return smells
}

func (c *CleanupEngine) Clean(code string, lang Language) (string, []*CodeSmell, error) {
	smells := c.DetectSmells(code, lang)
	if len(smells) == 0 {
		return code, nil, nil
	}

	smellsDescription := make([]string, len(smells))
	for i, s := range smells {
		smellsDescription[i] = fmt.Sprintf("- %s: %s", s.Name, s.Description)
	}
	smellsDescriptionStr := strings.Join(smellsDescription, "\n")

	prompt := fmt.Sprintf(`You are an expert Refactor Bot. Clean up the following %s code to be fully idiomatic.

WE DETECTED THESE MIGRATION SMELLS:
%s

CODE TO FIX:
%s

RULES:
- FIX ALL detected smells.
- Use standard idiomatic patterns (e.g., @dataclass, typing, proper serialization).
- Output ONLY the fixed code, no explanations or markdown fences.

`, lang, smellsDescriptionStr, code)

	cleaned, err := c.Model.Complete(prompt)
	if err != nil {
		return "", nil, err
	}
	cleaned = strings.TrimSpace(cleaned)
	if strings.HasPrefix(cleaned, "```") {
		cleaned = cleaned[3:]
	}

	return cleaned, smells, nil
}

func (c *CleanupEngine) stripMarkdown(s string) string {
	var out bytes.Buffer
	if strings.HasPrefix(s, "```") {
		out.WriteString(strings.TrimSpace(s[3:]))
		out.WriteString("\n")
	}
	return out.String()
}

const (
	Python = iota
)

File: core/src/migrate/clean.rs
use anyhow::Result;
use serde::{Deserialize, Serialize};
use crate::model::CodexModel;
use super::detect::Language;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodeSmell {
    pub name: String,
    pub description: String,
    pub line_hint: Option<usize>,
}

pub struct CleanupEngine<'a> {
    model: &'a (dyn CodexModel + Send + Sync),
}

impl<'a> CleanupEngine<'a> {
    pub fn new(model: &'a (dyn CodexModel + Send + Sync)) -> Self {
        Self { model }
    }

    pub fn detect_smells(&self, code: &str, lang: Language) -> Vec<CodeSmell> {
        let mut smells = Vec::new();

        match lang {
            Language::Python => {
                // Smell: JSON dump of objects
                if code.contains("json.dump(") && !code.contains("asdict(") && !code.contains("JSONEncoder") {
                    smells.push(CodeSmell {
                        name: "Custom Object Serialization".to_string(),
                        description: "Found `json.dump` without `asdict` or custom encoder. This will crash for custom classes.".to_string(),
                        line_hint: None,
                    });
                }
                // Smell: Missing dataclasses
                if code.contains("def __init__(self") && !code.contains("@dataclass") {
                    smells.push(CodeSmell {
                        name: "Missing Dataclasses".to_string(),
                        description: "Boilerplate __init__ mirrors source struct; consider using @dataclass.".to_string(),
                        line_hint: None,
                    });
                }
                // Smell: Static methods for defaults
                if code.contains("@staticmethod") && (code.contains("default") || code.contains("max_entries")) {
                    smells.push(CodeSmell {
                        name: "Rust-style Static Helpers".to_string(),
                        description: "Static methods like `default_max_entries` should be converted to default parameters.".to_string(),
                        line_hint: None,
                    });
                }
            }
            _ => {}
        }

        smells
    }

    pub fn clean(&self, code: &str, lang: Language) -> Result<(String, Vec<CodeSmell>)> {
        let smells = self.detect_smells(code, lang);
        if smells.is_empty() {
            return Ok((code.to_string(), Vec::new()));
        }

        let smells_desc = smells.iter()
            .map(|s| format!("- {}: {}", s.name, s.description))
            .collect::<Vec<_>>()
            .join("\n");

        let prompt = format!(
            "You are an expert Refactor Bot. Clean up the following {} code to be fully idiomatic.\n\n\
             WE DETECTED THESE MIGRATION SMELLS:\n\
             {}\n\n\
             CODE TO FIX:\n\
             {}\n\n\
             RULES:\n\
             - FIX ALL detected smells.\n\
             - Use standard idiomatic patterns (e.g., @dataclass, typing, proper serialization).\n\
             - Output ONLY the fixed code, no explanations or markdown fences.\n",
            lang, smells_desc, code
        );

        let cleaned = self.model.complete(&prompt)?;
        let cleaned = self.strip_markdown(&cleaned);

        Ok((cleaned, smells))
    }

    fn strip_markdown(&self, s: &str) -> String {
        let mut out = s.to_string();
        if out.starts_with("```") {
            if let Some(first_newline) = out.find('\n') {
                out = out[first_newline + 1..].to_string();
            }
            if out.ends_with("```") {
                out = out[..out.len() - 3].to_string();
            }
        }
        out.trim().to_string()
    }
}
